buildkit/internal/transform/transforms.go

```diff
@@ -60,66 +62,104 @@ type virtualEnv struct {
 	RequirementsFile string
 }
 
-func newTransformer() *Transformer {
+func newTransformer(buildArgs map[string]string) *Transformer {
 	return &Transformer{
+		buildArgs:      buildArgs,
 		pythonVersions: map[string]struct{}{},
 	}
 }
 
 // Transform processes the custom dockerfile and extracts the first FROM and converts the rest of the user dockerfile
 // into standard Docker commands
 func Transform(dockerFile []byte, buildArgs map[string]string) (*parser.Node, *parser.Node, error) {
-	ast, err := dockerfile.Parse(dockerFile)
-	if err != nil {
-		return nil, nil, err
-	}
-	transformedAst := &parser.Node{StartLine: -1}
+	return newTransformer(buildArgs).Transform(dockerFile)
+}
+
+func (transformer *Transformer) parsePreamble(tokens chan *parser.Node) (*parser.Node, int, error) {
 	preamble := &parser.Node{StartLine: -1}
-	transformer := newTransformer()
-	inPreamble := true
 	adjustedLine := 0
-	for _, node := range ast.Children {
-		if inPreamble {
-			switch strings.ToUpper(node.Value) {
-			case argCommand:
-				if err = processArg(node, buildArgs); err != nil {
-					return nil, nil, err
-				}
-			case fromCommand:
-				// finish preamble when we encounter the first FROM
-
-				// Try to use the astro-runtime base (non-onbuild) image. If we have any error, "fail safe"
-				// and leave it unmodified.
-				_ = ensureValidBaseImage(node, buildArgs)
-				adjustedLine = -node.EndLine
-				inPreamble = false
-			case pyenvCommand:
-				return nil, nil, fmt.Errorf("%s cannot appear before the first FROM", pyenvCommand)
+
+tokens:
+	for node := range tokens {
+		preamble.AddChild(node, node.StartLine, node.EndLine)
+		switch strings.ToUpper(node.Value) {
+		case argCommand:
+			if err := transformer.processArg(node); err != nil {
+				return nil, 0, err
 			}
-			preamble.AddChild(node, node.StartLine, node.EndLine)
-			continue
+		case fromCommand:
+			// finish preamble when we encounter the first FROM
+
+			// Try to use the astro-runtime base (non-onbuild) image. If we have any error, "fail safe"
+			// and leave it unmodified.
+			_ = transformer.ensureValidBaseImage(node)
+			adjustedLine = -node.EndLine
+			break tokens
+		case pyenvCommand:
+			return nil, 0, fmt.Errorf("%s cannot appear before the first FROM", pyenvCommand)
 		}
+	}
+	return preamble, adjustedLine, nil
+}
+
+func (transformer *Transformer) parseMain(tokens chan *parser.Node, lineNoAdjustment int) (*parser.Node, error) {
```

Comment:
```suggestion
func (transformer *Transformer) parseMain(tokens chan *parser.Node, lineNumAdjustment int) (*parser.Node, error) {
```
Just to be clear it doesn't mean no adjustment